### DISCUSSION TITLE
Count the total number of successful steps in a job using github actions

### DISCUSSION THREAD

**Discussion by uditgaurav** [community/community#25362] [Actions]
<p>Hi all,<br>
This is for getting the total number of successful steps in a Github action job.<br>
I need this value for further decision making.<br>
Example:</p>
<pre><code class="lang-auto">name: Litmus-CI
on:
  issue_comment:
    types: [created, edited]

jobs:
  job1:
    if: contains(github.event.comment.html_url, '/pull/') &amp;&amp; startsWith(github.event.comment.body, '/run-e2e')
    runs-on: ubuntu-latest
    steps:

   ...step1...

   ...step2...
  
   ...step3...

   ...step4...

   ...step5...

</code></pre>
<p>Let us suppose I want to run step5 only if anyone of step4 or step3 ran successfully(or both doesn’t skip) how to build such logic around this?</p>
<p>If I have some parameters to get the total number of steps ran successfully then this would be done easily.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/uditgaurav">@uditgaurav</a>,</p>
<aside class="quote no-group" data-username="uditgaurav" data-post="1" data-topic="116898" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 uditgaurav:</div>
<blockquote>
<p>Let us suppose I want to run step5 only if anyone of step4 or step3 ran successfully(or both doesn’t skip) how to build such logic around this?</p>
</blockquote>
</aside>
<p>Step status can be got from step context: ‘steps.step_id.outcome’ or ‘steps.step_id.conclusion’. Please check the official <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#steps-context" rel="nofollow noopener">doc</a> for more details.<br>
You can use a combined expression for step5, code sample as below:</p>
<pre><code class="lang-auto">      - name: step3
        id: step3
        continue-on-error: true
        run: |
          exit 1
      - name: step4
        id: step4
        continue-on-error: true
        run: |
          exit 1
      - name: step5
        if: always()&amp;&amp;(steps.step3.outcome=='success' || steps.step4.outcome=='success') # step5 will be skipped if both step3 and step4 are not successful.
        run: |
          echo ${{ steps.step3.outcome }}
          echo ${{ steps.step4.outcome }}
</code></pre>
<p>Please change the outcome value according to your requirements.</p>
<aside class="quote no-group" data-username="uditgaurav" data-post="1" data-topic="116898" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 uditgaurav:</div>
<blockquote>
<p>If I have some parameters to get the total number of steps ran successfully then this would be done easily.</p>
</blockquote>
</aside>
<p>You can check the step context at the last of the job.</p>
<pre><code class="lang-auto">      - name: Dump steps context
        env:
          STEPS_CONTEXT: ${{ toJson(steps) }}
        run: echo "$STEPS_CONTEXT"
</code></pre>
<p>My output is as below, you can parse the context of outcome value, and caculate the numbe of successful steps in the job.<br>
<img src="https://user-images.githubusercontent.com/52772761/181090609-de285e36-5208-4651-818e-a8b1324f382b.png" alt="stepcontext" data-base62-sha1="x6MlYETiw0FeZbzDSriWpW7zvDc" width="581" height="477"></p>
<p>Thanks.</p>

**Comment by blakethepatton** [comment]
<p>Hi <a class="mention" href="/weide-zhou">@weide-zhou</a>, I’m running my job on push and have only been seeing an empty object <code>{}</code> returned when I encode the steps object. Not entirely sure what I might be doing wrong, my action is the copy pasted contents of: <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#example-printing-context-information-to-the-log-file" rel="nofollow noopener">https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#example-printing-context-information-to-the-log-file</a></p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/blakethepatton">@blakethepatton</a>,</p>
<p>Please confirm that you add ‘id’ for your steps.</p>
<pre><code class="lang-auto">      - name: step4
        id: step4     # add step id here!
        continue-on-error: true
        run: |
          exit 1
      - name: Dump steps context
        env:
          STEPS_CONTEXT: ${{ toJson(steps) }}
        run: echo "$STEPS_CONTEXT"
</code></pre>
<p>If there still no output, please share your yaml for further investigation.</p>
<p>Thanks.</p>

**Comment by blakethepatton** [comment]
<p>Yes <a class="mention" href="/weide-zhou">@weide-zhou</a>, adding id made the steps visible. Thank you so much I’ve spent probably three hours now trying to figure that out! What is the purpose of ID verses name? Other than id implies that it will be unique to that step?</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/blakethepatton">@blakethepatton</a>,</p>
<p>step id is a unique identifier for the step. You can use the  <code>id</code>  to reference the step in contexts. It’s mentioned in the doc <a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsid" rel="nofollow noopener">here</a>.</p>
<p>Thanks</p>

**Comment by blakethepatton** [comment]
<p>Is there a way to access steps from inside a javascript runner without having to explicitly pass it in through an input variable? Additionally, is there a way to get the message behind why a step failed? Or would you need to redirect the console output of the command you’re running into a output variable and then access the error that way.</p>
